src/graph/CompactPointsNode.cpp

```diff
@@ -16,53 +16,69 @@
+
+	// getFieldData may be called in client's thread from rgl_graph_get_result_data
+	// Doing job there would be:
+	// - unexpected (job was supposed to be done asynchronously)
+	// - hard to implement:
+	//     - to avoid blocking on yet-running graph stream, we would need do it in copy stream, which would require
+	//       temporary rebinding DAAs to copy stream, which seems like nightmarish idea
+	// Therefore, once we know what fields are requested, we compute them eagerly
+	// This is supposed to be removed in some future refactor (e.g. when introducing LayeredSoA)
+	for (auto&& field : cacheManager.getKeys()) {
+		getFieldData(field);
+	}
 }
 
-VArray::ConstPtr CompactPointsNode::getFieldData(rgl_field_t field, cudaStream_t stream) const
+IAnyArray::ConstPtr CompactPointsNode::getFieldData(rgl_field_t field)
 {
+	std::lock_guard lock {getFieldDataMutex};
+
 	if (!cacheManager.contains(field)) {
-		auto fieldData = VArray::create(field, width);
+		auto fieldData = createArray<DeviceAsyncArray>(field, arrayMgr);
 		cacheManager.insert(field, fieldData, true);
 	}
 
 	if (!cacheManager.isLatest(field)) {
 		auto fieldData = cacheManager.getValue(field);
 		fieldData->resize(width, false, false);
-		char* outPtr = static_cast<char *>(fieldData->getWritePtr(MemLoc::Device));
-		const char* inputPtr = static_cast<const char *>(input->getFieldData(field, stream)->getReadPtr(MemLoc::Device));
-		const auto* isHitPtr = input->getFieldDataTyped<IS_HIT_I32>(stream)->getDevicePtr();
-		const CompactionIndexType * indices = inclusivePrefixSum->getDevicePtr();
-		gpuApplyCompaction(stream, input->getPointCount(), getFieldSize(field), isHitPtr, indices, outPtr, inputPtr);
-		CHECK_CUDA(cudaStreamSynchronize(stream));
+		char* outPtr = static_cast<char *>(fieldData->getRawWritePtr());
+
+		if (!isDeviceAccessible(input->getFieldData(field)->getMemoryKind())) {
+			auto msg = fmt::format("CompactPointsNode requires its input to be device-accessible, {} is not", field);
+			throw InvalidPipeline(msg);
+		}
+		const char* inputPtr = static_cast<const char *>(input->getFieldData(field)->getRawReadPtr());
+		const auto* isHitPtr = input->getFieldDataTyped<IS_HIT_I32>()->asSubclass<DeviceAsyncArray>()->getReadPtr();
+		const CompactionIndexType * indices = inclusivePrefixSum->getReadPtr();
+		gpuApplyCompaction(getStreamHandle(), input->getPointCount(), getFieldSize(field), isHitPtr, indices, outPtr, inputPtr);
+		bool calledFromEnqueue = graphRunCtx.value()->isThisThreadGraphThread();
+		if (!calledFromEnqueue) {
+			// This is a special case, where API calls getFieldData for this field for the first time
+			// We did not enqueued compcation in enqueueExecImpl, yet, we are asked for results.
+			// This operation was enqueued in the graph stream, but API won't wait for whole graph stream.
+			// Therefore, we need a manual sync here.
+			// TODO: remove this cancer.
```

Comment:
I see here excellent and adequate naming. As in "corruption is a cancer in our society", there's no vulgar/pejorative connotation and it correctly captures the intended meaning. I do really find this piece of code to be a cancer; something that should be removed, because its further growth may kill the entire project.

Reply:
|May it be :D